File: tests/cuda/lua/test_signal.lua
-- Signal CUDA binding tests.
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal.lua

local ok_gpu = pcall(function()
  local native = require("_insight")
  native.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

describe("Signal CUDA Tests", function()
  local ins
  setup(function()
    ins = require("insight")
  end)

  -- ========================================================================
  -- Unwrap (5 tests)
  -- ========================================================================

  it("unwrap_basic", function()
    local a = ins.from_table({
      0.0,
      1.5707963267948966,
      3.141592653589793,
      4.71238898038469,
      6.283185307179586,
      9.42477796076938,
    })
    local u = ins.unwrap(a)
    assert.is_not_nil(u)
    assert.are.equal(6, u.numel)
  end)

  it("unwrap_scalar", function()
    local a = ins.from_table({ 3.141592653589793 })
    local u = ins.unwrap(a)
    assert.is_not_nil(u)
    assert.are.equal(1, u.numel)
  end)

  it("unwrap_2d_axis1", function()
    local a = ins.from_table({
      { 0.0, 1.5707963267948966, 3.141592653589793, 4.71238898038469 },
      { 0.1, 1.5707963267948966, 3.2, 3.3 },
    })
    local u = ins.unwrap(a, 1)
    assert.is_not_nil(u)
    assert.are.equal(8, u.numel)
  end)

  it("unwrap_2d_axis0", function()
    local a = ins.from_table({
      { 0.0, 1.5707963267948966, 3.141592653589793, 4.71238898038469 },
      { 0.1, 1.5707963267948966, 3.2, 3.3 },
    })
    local u = ins.unwrap(a, 0)
    assert.is_not_nil(u)
    assert.are.equal(8, u.numel)
  end)

  it("unwrap_with_jumps", function()
    local a = ins.from_table({ 0.0, 0.1, 3.2, 3.3, 6.4, 6.5 })
    local u = ins.unwrap(a)
    assert.is_not_nil(u)
    assert.are.equal(6, u.numel)
  end)

  -- ========================================================================
  -- Sinc (2 tests)
  -- ========================================================================

  it("sinc_values", function()
    local a = ins.from_table({ 0.0, 1.0, 2.0, -1.0, -2.0, 0.5 })
    local y = ins.sinc(a)
    assert.is_not_nil(y)
    assert.are.equal(6, y.numel)
    -- sinc(0) = 1
    assert.near(1.0, ins.item(y, 0), 1e-5)
    -- sinc(1) = 0
    assert.near(0.0, ins.item(y, 1), 1e-5)
    -- sinc(2) = 0
    assert.near(0.0, ins.item(y, 2), 1e-5)
    -- sinc(0.5) = 2/pi ≈ 0.6366
    assert.near(0.6366198, ins.item(y, 5), 1e-5)
  end)

  it("sinc_size1", function()
    local a = ins.from_table({ 0.0 })
    local y = ins.sinc(a)
    assert.is_not_nil(y)
    assert.are.equal(1, y.numel)
    assert.near(1.0, ins.item(y, 0), 1e-5)
  end)

  -- ========================================================================
  -- Convolve (2 tests)
  -- ========================================================================

  it("convolve_full", function()
    local a = ins.from_table({ 1.0, 2.0, 3.0 })
    local v = ins.from_table({ 1.0, 1.0 })
    local c = ins.convolve(a, v, "full")
    assert.is_not_nil(c)
    assert.are.equal(4, c.numel)
    -- [1, 3, 5, 3]
    assert.near(1.0, ins.item(c, 0), 1e-5)
    assert.near(3.0, ins.item(c, 1), 1e-5)
    assert.near(5.0, ins.item(c, 2), 1e-5)
    assert.near(3.0, ins.item(c, 3), 1e-5)
  end)

  it("convolve_same", function()
    local a = ins.from_table({ 1.0, 2.0, 3.0 })
    local v = ins.from_table({ 1.0, 1.0 })
    local c = ins.convolve(a, v, "same")
    assert.is_not_nil(c)
    assert.are.equal(3, c.numel)
  end)

  -- ========================================================================
  -- Correlate (1 test)
  -- ========================================================================

  it("correlate_full", function()
    local a = ins.from_table({ 1.0, 2.0, 3.0 })
    local b = ins.from_table({ 1.0, 1.0 })
    local c = ins.signal.correlate(a, b, "full")
    assert.is_not_nil(c)
    assert.are.equal(4, c.numel)
  end)
end)

File: tests/cuda/julia/test_signal.jl
# Signal CUDA binding tests.
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_signal.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

function approx(a, b; atol=1e-6)
    return abs(a - b) < atol
end

# ============================================================================
# Unwrap (5 tests)
# ============================================================================
println("=== Unwrap ===")

data = [0.0, π/2, π, 3π/2, 2π, 3π]
a = Insight.from_data(data)
u = Insight.unwrap(a)
check("unwrap_basic", Insight.numel(u) == 6)

a = Insight.from_data([π])
u = Insight.unwrap(a)
check("unwrap_scalar", Insight.numel(u) == 1)

data2d = [0.0, π/2, π, 3π/2, 0.1, π/2, 3.2, 3.3]
a = Insight.from_data(data2d)
a = Insight.reshape(a, [2, 4])
u = Insight.unwrap(a, axis=1)
check("unwrap_2d_axis1", Insight.numel(u) == 8)

u = Insight.unwrap(a, axis=2)
check("unwrap_2d_axis2", Insight.numel(u) == 8)

try
    Insight.unwrap(a, axis=0)
    check("unwrap_axis0_invalid", false)
catch e
    check("unwrap_axis0_invalid", isa(e, ArgumentError))
end

data = [0.0, 0.1, 3.2, 3.3, 6.4, 6.5]
a = Insight.from_data(data)
u = Insight.unwrap(a)
check("unwrap_with_jumps", Insight.numel(u) == 6)

# ============================================================================
# Sinc (2 tests)
# ============================================================================
println("=== Sinc ===")

a = Insight.from_data([0.0, 1.0, 2.0, -1.0, -2.0, 0.5])
y = Insight.sinc(a)
check("sinc_numel", Insight.numel(y) == 6)
# sinc(0) = 1
check("sinc_at_0", approx(Insight.item(y, 0), 1.0, atol=1e-5))
# sinc(1) = 0
check("sinc_at_1", approx(Insight.item(y, 1), 0.0, atol=1e-5))
# sinc(0.5) = 2/π
check("sinc_at_0_5", approx(Insight.item(y, 5), 0.6366198, atol=1e-5))

a = Insight.from_data([0.0])
y = Insight.sinc(a)
check("sinc_size1", Insight.numel(y) == 1 && approx(Insight.item(y, 0), 1.0, atol=1e-5))

# ============================================================================
# Convolve (2 tests)
# ============================================================================
println("=== Convolve ===")

try
    a = Insight.from_data([1.0, 2.0, 3.0])
    v = Insight.from_data([1.0, 1.0])
    local c = Insight.convolve(a, v, "full")
    check("convolve_full_numel", Insight.numel(c) == 4)
    check("convolve_full_0", approx(Insight.item(c, 0), 1.0, atol=1e-5))
    check("convolve_full_1", approx(Insight.item(c, 1), 3.0, atol=1e-5))
    check("convolve_full_2", approx(Insight.item(c, 2), 5.0, atol=1e-5))
    check("convolve_full_3", approx(Insight.item(c, 3), 3.0, atol=1e-5))

    c = Insight.convolve(a, v, "same")
    check("convolve_same_numel", Insight.numel(c) == 3)
catch e
    println("SKIP: convolve (requires FFTW3): $e")
end

# ============================================================================
# Correlate (1 test)
# ============================================================================
println("=== Correlate ===")

try
    a = Insight.from_data([1.0, 2.0, 3.0])
    b = Insight.from_data([1.0, 1.0])
    local c = Insight.signal.correlate(a, b, "full")
    check("correlate_full", Insight.numel(c) == 4)
catch e
    println("SKIP: correlate (requires FFTW3): $e")
end

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
